src/dask_awkward/lib/io/columnar.py

```diff
@@ -74,105 +55,3 @@ def mock_empty(self: S, backend: BackendT = "cpu") -> AwkwardArray:
                 highlevel=True,
             ),
         )
-
-    def prepare_for_projection(
-        self: S,
-    ) -> tuple[AwkwardArray, TypeTracerReport, FormStructure]:
-        form = form_with_unique_keys(self.form, "@")
-
-        # Build typetracer and associated report object
-        (meta, report) = typetracer_with_report(
-            form,
-            highlevel=True,
-            behavior=self.behavior,
-            buffer_key=render_buffer_key,
-        )
-
-        return (
-            cast(AwkwardArray, meta),
-            report,
-            trace_form_structure(form, buffer_key=render_buffer_key),
-        )
-
-    def necessary_columns(
-        self: S,
-        report: TypeTracerReport,
-        state: FormStructure,
-    ) -> frozenset[str]:
-        ## Read from stash
-        # Form hierarchy information
-        form_key_to_parent_form_key = state["form_key_to_parent_form_key"]
-        form_key_to_child_form_keys: dict[str, list[str]] = {}
-        for child_key, parent_key in form_key_to_parent_form_key.items():
-            form_key_to_child_form_keys.setdefault(parent_key, []).append(child_key)  # type: ignore
-        form_key_to_form = state["form_key_to_form"]
-        # Buffer hierarchy information
-        form_key_to_buffer_keys = state["form_key_to_buffer_keys"]
-        # Column hierarchy information
-        form_key_to_path = state["form_key_to_path"]
-
-        # Require the data of metadata buffers above shape-only requests
-        data_buffers = {
-            *report.data_touched,
-            *buffer_keys_required_to_compute_shapes(
-                parse_buffer_key,
-                report.shape_touched,
-                form_key_to_parent_form_key,
-                form_key_to_buffer_keys,
-            ),
-        }
-
-        # We can't read buffers directly, but if we encounter a metadata
-        # buffer, then we should be able to pick any child.
-        paths = set()
-        wildcard_form_key = set()
-        for buffer_key in data_buffers:
-            form_key, attribute = parse_buffer_key(buffer_key)
-            if attribute in METADATA_ATTRIBUTES:
-                wildcard_form_key.add(form_key)
-            else:
-                paths.add(form_key_to_path[form_key])
-
-        # Select the most appropriate column for each wildcard
```

Comment:
@martindurant is the plan to restore e.g. wildcard matching in this PR?

Reply:
I don't think that happens in this PR, this one's just about auto-selecting required columns. We might need more work for the case where the shape of something is required but not the data.

Reply:
I'm only surface-glancing, but I noted that this logic has been removed and didn't see anywhere that it was recovered. Might have missed something!

Reply:
You are right, this should live after checking shape_touched_in, more complex than I had hoped.

Reply:
OK< I have some shonky code in optimize.optimize_columns to deal with shapes. I am probably still missing some touch-buffers on the outputs, but most things are in line. 

It's all pretty ugly and in the latest commit I seem to have introduced some typetracer->numpy coersion.